### PR TITLE
[#15] [UI] As an anonymous user, I can see the forgot password screen

### LINKED ISSUE #15: [UI] As an anonymous user, I can see the forgot password screen
## Why

The user may forget his/her password after a long time of using the application. To provide the user a method to get back his/her password, the application requires the user to fill registered email.

## Acceptance criteria

After the user taps on the `Forgot?` button on the login screen, navigate to the forgot password screen.

In the forgot password screen, show:

- Nimble company logo
- Forgot password instruction
- A text field for inputting registered email
- A `Reset` button

It is recommended to have a format verification for the email input

## Design

- [Login screen](https://www.figma.com/proto/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=31%3A195&scaling=scale-down)
- [Password screen](https://www.figma.com/proto/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=31%3A457&scaling=scale-down)

### PR DESCRIPTION
**Pull request by nmint8m**
- Close #15

## What happened 👀

The user may forget his/her password after a long time of using the application. To provide the user a method to get back his/her password, the application requires the user to fill registered email.

After the user taps on the `Forgot?` button on the login screen, navigate to the forgot password screen.

In the forgot password screen, show:

- Nimble company logo
- Forgot password instruction
- A text field for inputting registered email
- A `Reset` button

It is recommended to have a format verification for the email input.

### CODE REVIEW

File: lib/views/forget_password/forget_password_form.dart [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        borderSide: BorderSide.none,
+        borderRadius: BorderRadiuses.circular12,
+      ),
+      fillColor: Colors.white24,
+      filled: true,
+      contentPadding: EdgeInsets.only(
+        top: 18.0,
+        bottom: 18.0,
+        left: 12.0,
+        right: rightPadding,
```
Comment (manh-t) [inline lib/views/forget_password/forget_password_form.dart]: we only have an input field in this screen. I think we can put `rightPadding` directly here 🙋

File: lib/views/forget_password/forget_password_form.dart [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      return;
+    }
+    Keyboard.hideKeyboard(context);
+    // TODO: - Call API
+  }
+
+  String? _validateEmailMessage(String? email) {
+    ref.read(forgetPasswordViewModelProvider.notifier).validateEmail(email);
+    return ref
+        .read(forgetPasswordViewModelProvider.notifier)
```
Comment (manh-t) [inline lib/views/forget_password/forget_password_form.dart]: extract `ref.read(forgetPasswordViewModelProvider.notifier)` to a variable to reuse would be nice

Comment (manh-t) [reply lib/views/forget_password/forget_password_form.dart]: @nmint8m also, I would recommend creating a habit to create the callback variable to handle the actions related to ViewModel. Its benefit can be:
- The widget can be reusable on multiple screens ( we don't stick a widget to a specific screen)
- All ViewModel's methods will be called in the View layer, so later we don't need to visit the nested widget to find the VM's methods. It's easy to maintain. 😄

File: lib/views/forget_password/forget_password_view.dart [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        fit: BoxFit.cover,
+        alignment: Alignment.center,
+      );
+
+  Widget get _linearGradientBlurBackground => Stack(
+        children: [
+          _backgroundImage,
+          ImageFiltered(
+            imageFilter: ImageFilter.blur(sigmaX: 10, sigmaY: 10),
+            child: ShaderMask(
```
Comment (manh-t) [inline lib/views/forget_password/forget_password_view.dart]: 👍

File: lib/views/forget_password/forget_password_view.dart [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  Widget build(BuildContext context) {
+    _setupStateListener();
+    return _defaultForgetPasswordView();
+  }
+
+  Widget _defaultForgetPasswordView() {
+    return Stack(
+      children: [
+        _background,
+        Scaffold(
```
Comment (manh-t) [inline lib/views/forget_password/forget_password_view.dart:89]: `Scaffold` should be the root widget imo

Comment (nmint8m) [reply lib/views/forget_password/forget_password_view.dart:89]: From my perspective, I think the `Scaffold` is the widget that can divide the screen into basic components, such as app bar, body, bottom bar, etc. It's not forced to be the root widget.

Actually, I tried the approach of having the `Scaffold` as the root widget for this screen. The problem is that I just can put the background widget into the `body` of the `Scaffold`, so the background image cannot cover the whole page:

```dart
Widget _defaultForgetPasswordView() {
    return Stack(
      children: [
        Scaffold(
          backgroundColor: Colors.transparent,
          appBar: _appBar,
          body: Container(
            decoration: BoxDecoration(
              image: DecorationImage(image: _backgroundImage.image),
            ),
            child: BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
              child: Container(
                color: Colors.black45,
                child: SafeArea(
                  child: Container(
                    padding: const EdgeInsets.symmetric(
                        vertical: 30, horizontal: 24),
                    child: Column(
                      children: [
                        const SizedBox(height: 60),
                        _logo,
                        const SizedBox(height: 20),
                        _intruction(context),
                        const SizedBox(height: 100),
                        _form,
                      ],
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
```

Screenshot 2023-02-13 at 17 21 28

Anyway, I think we can try different approaches to implement one UI. It's nice.

File: lib/views/forget_password/forget_password_form.dart [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        children: [
+          _emailTextField,
+          const SizedBox(height: 20),
+          _resetButton,
+        ],
+      ),
+    );
+  }
+
+  Future<void> _reset() async {
```
Comment (doannimble) [inline lib/views/forget_password/forget_password_form.dart]:
```suggestion
  Future<void> _resetForm() async {
```

File: lib/views/forget_password/forget_password_view.dart [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  void _setupStateListener() {
+    ref.listen<ForgetPasswordState>(forgetPasswordViewModelProvider,
+        (_, state) {
+      context.showOrHideLoadingIndicator(
+        shouldShow: state == const ForgetPasswordState.loading(),
+      );
+      state.maybeWhen(
+        error: (error) {
+          ScaffoldMessenger.of(context).showSnackBar(
```
Comment (doannimble) [inline lib/views/forget_password/forget_password_view.dart]: @nmint8m I think we should put `showSnackBar` in `ContextExt` for convenience using. What do you think?

File: lib/views/forget_password/forget_password_view.dart [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        child: FittedBox(
+          fit: BoxFit.contain,
+          child: Image(
+            image: Assets.images.nimbleLogoWhite.image().image,
+            fit: BoxFit.scaleDown,
+          ),
+        ),
+      );
+
+  Widget _intruction(BuildContext context) => Text(
```
Comment (doannimble) [inline lib/views/forget_password/forget_password_view.dart]:
```suggestion
  Widget _instruction(BuildContext context) => Text(
```

File: lib/views/forget_password/forget_password_view.dart [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        ),
+      );
+
+  Widget get _backgroundImage => Image(
+        image: Assets.images.nimbleBackground.image().image,
+        fit: BoxFit.cover,
+        alignment: Alignment.center,
+      );
+
+  Widget get _linearGradientBlurBackground => Stack(
```
Comment (doannimble) [inline lib/views/forget_password/forget_password_view.dart]: @nmint8m I think we can define this as a common widget for reusable (login/forget password using the same?) 🤔

**Review by doannimble — COMMENTED** [review]
rest lgtm

File: lib/views/forget_password/forget_password_view.dart (hunk 4 above)
Comment (manh-t) [reply lib/views/forget_password/forget_password_view.dart:89]: ah I see, you're right. It's the component to divide the screen into basic components 😄 anw, let's keep it now

**Review by manh-t — APPROVED** [review]

**Pull request merged**